20241016-R-error-code-metrics.md

```diff
@@ -0,0 +1,111 @@
+
+Initial Error Codes Proposal: [proposals/#21](https://github.com/dapr/proposals/pull/21)
+
+Error Codes Implementation Parent: [dapr/#6068](https://github.com/dapr/dapr/issues/6068)
+
+## Expectations and alternatives
+
+### Expectations
+- When opted-in, it would add up to  `N * app`  (N = number of error codes) metrics for scraping, today that's around `52 * app`.
+- If a component-based error, codes could repeat for multiple components used by the same app, and wouldn't be differentiated by the metrics, which only provide the `app-id`.
+- Many logged error codes are currently defined in [pkg/messages/predefined.go](https://github.com/dapr/dapr/blob/master/pkg/messages/predefined.go), but not all. Implementation requires case-by-case analysis of usage of error codes in that uppercase format `ERR_NAME_TAG`.
+
+### Alternatives
+- One alternative solution which retains error codes as logs only would be to change the level of logs containing these error codes from `debug` to `info`. However, this increases noise and raises logs which may be unhelpful at `info` level.
+- Another alternative solution would be to log error codes separately with opt-in functionality, but this achieves the same result as adding metrics, but increase log noise and doesn't introduce a simple way to actually work with the errors.
+
+## Implementation Details
+
+### Solution
+
+The feature is enabled via an opt-in flag within Dapr HTTP metrics (disabled by default). The goal of the opt-in functionality is to prevent the increased cost of additional metrics if they aren't being used.
+
+```yaml
+spec:
+  metric:
+    enabled: true
+	recordErrorCodes: true # <-- new opt-in flag
+	http:
+		...
+	...
+...
+```
+From the user's perspective, the metric `dapr_error_code_count` could be split by `app-id` and `error_code` like so:
+```go
+dapr_error_code_count{app_id="service-a", error_code="ERR_ACTOR_RUNTIME_NOT_FOUND"} 4
+dapr_error_code_count{app_id="service-a", error_code="ERR_ACTOR_INVOKE_METHOD"} 12
+dapr_error_code_count{app_id="service-b", error_code="ERR_ACTOR_RUNTIME_NOT_FOUND"} 55
+dapr_error_code_count{app_id="service-b", error_code="ERR_DIRECT_INVOKE"} 2
+```
+
+Error codes are currently defined as a struct, with the specific code as `APIError.tag`:
+```go
+ErrHealthNotReady  =  APIError{message: "dapr is not ready",  tag: "ERR_HEALTH_NOT_READY", httpCode: http.StatusInternalServerError,  grpcCode: grpcCodes.Internal}
+```
+
+...and are currently implemented case-by-case across Dapr, but generally as-is with `log.Debug(msg)`. Metric recording would occur in tandem with the log statement:
+```go
+func  (a *api)  onGetHealthz(w http.ResponseWriter, r *http.Request)  {
+	if !a.readyStatus {
+		msg := messages.ErrHealthNotReady
+		respondWithError(w, msg)
+		log.Debug(msg)
+		ErrorCodeMetrics.RecordError(msg.tag)
+		return
+	}
+...
+}
+```
+
+With `RecordError()` simply recording OpenCensus metrics like other metric recorders found in `pkg/diagnostics`:
```

Comment:
Just a note that we're moving to Open Telemetry, so the specific code in the example here will change.